### PR TITLE
refactor(select): Fix all syncing issues with the Select component

### PR DESCRIPTION
**Pull request by NicholasBoll**
## Summary

The `useComboboxInputConstrained` hook was created for Comboboxes that can only have a value from within a set of defined options. The value is constrained by the options passed to it. `useComboboxInputConstrained` assumes there will be multiple inputs where one is presented to the user and the other is for the form/server. The hook keeps the model in sync with external ways to manipulate the input: React `value` prop ([controlled inputs](https://react.dev/reference/react-dom/components/input#controlling-an-input-with-a-state-variable)), browser/extension autofill, and testing tools.

`Select` was updated to use `useComboboxInputConstrained` which fixes several bugs at once:
- Fixes https://github.com/Workday/canvas-kit/issues/2888
- Fixes https://github.com/Workday/canvas-kit/issues/2675
- Fixes https://github.com/Workday/canvas-kit/issues/2616
- Fixes https://github.com/Workday/canvas-kit/issues/2533

## Release Category
Components

---

## Checklist

- [x] MDX documentation adheres to Canvas Kit's [Documentation Guidelines](https://workday.github.io/canvas-kit/?path=/docs/guides-documentation-guidelines--page)
- [x] Label `ready for review` has been added to PR

## For the Reviewer

- [x] PR title is short and descriptive
- [x] PR summary describes the change (Fixes/Resolves linked correctly)

## Areas for Feedback? (optional)

- [ ] Code
- [ ] Documentation
- [ ] Testing
- [ ] Codemods

### CODE REVIEW

File: modules/react/combobox/lib/hooks/useComboboxKeyboardTypeAhead.ts [hunk 1]
```diff
@@ -22,17 +22,17 @@ import {useComboboxModel} from './useComboboxModel';
  * ```
  */
 export const useComboboxKeyboardTypeAhead = createElemPropsHook(useComboboxModel)(model => {
-  const keySofar = React.useRef('');
+  const keySoFar = React.useRef('');
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/hooks/useComboboxKeyboardTypeAhead.ts:25]: I think the other spelling made more sense 😉

**Review by mannycarrera4 — APPROVED** [review]

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx [hunk 2]
```diff
@@ -252,6 +253,36 @@ The following changes have been made to the API:
...
+
+The `useComboboxInput` hook, and the `Combobox.Input` component used to automatically update the
+input when an option was selected. This functionality has been split between
+`useComboboxInputArbitrary` and `useComboboxInputConstrained` depending on the combobox's value
+type. The `useComboboxInput` had the arbitrary functionality built in. This has been separated out.
+The `<Select>` component has been updated to use `useComboboxInputConstrained` hook and the
+`Autocomplete` example uses the `useComboboxInputArbitrary` hook. This is a breaking change if you
+use the `Combobox.Input` component directly. You'll have to either pass the
+`useComboboxInputArbitrary` elemProps hook directly to the `<Combbox.Input>` or create a new
+component composing them.
```
Comment (josh-bagwell) [inline modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:268]:
```suggestion
The `useComboboxInput` hook, and the `Combobox.Input` component used to automatically update the
input when an option was selected. This functionality has been split between
`useComboboxInputArbitrary` and `useComboboxInputConstrained` depending on the combobox's value
type. The `useComboboxInput` had the arbitrary functionality built in but, now this has been separated out.
The `<Select>` component has been updated to use the `useComboboxInputConstrained` hook and the
`Autocomplete` example uses the `useComboboxInputArbitrary` hook. This is a breaking change if you
use the `Combobox.Input` component directly. You'll have to either pass the
`useComboboxInputArbitrary` elemProps hook directly to the `<Combbox.Input>` or create a new
component composing them.
```

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx [hunk 3]
```diff
@@ -252,6 +253,36 @@ The following changes have been made to the API:
...
+  // ...
+</Combobox>
+
+// v12
+<Combobox>
+  <Combobox.Input elemPropsHook={useComboboxInputArbitrary} />
+  // ...
+</Combobox>
+
+// Better - create a specialized component
```
Comment (josh-bagwell) [inline modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:283]: Is there more to go here?

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx [hunk 4]
```diff
@@ -504,6 +536,9 @@ const theme: PartialEmotionCanvasTheme = {
 </CanvasProvider>;
 ```
 
+`Select` has been refactored to use `useComboboxInputConstrained` to sync the model and the `input`
+element regardless of the source. This makes the `Select` be a controllable component.
```
Comment (josh-bagwell) [inline modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:540]:
```suggestion
`Select` has been refactored to use `useComboboxInputConstrained` to sync the model and the `input`
element regardless of the source. This makes the `Select` a controllable component.
```

File: modules/react/combobox/stories/Combobox.mdx [hunk 5]
```diff
@@ -19,6 +19,33 @@ Authoring Practices Guide (APG):
...
+  exampple, the `name` attribute will be passed here. The `ref` will pointed to this element.
+
+### Arbitrary
+
+An arbitrary combobox allows the user to enter any value. The list of options are presented as
+suggestions and selecting an option will prefill the combobox with the value of the option. The user
+is still allowed to modify the combobox even after an option is entered. With arbitrary comboboxes,
+there is only one `input` element. Arbitrary combobox inputs should use the
+[useComboboxInputArbitrary](#usecomboboxinputarbirary) hook.
+
```
Comment (josh-bagwell) [inline modules/react/combobox/stories/Combobox.mdx:48]:
```suggestion
A constrained combobox can only have a value from a set of values presented as options. The user can
pick from these values, but cannot input a value outside the options provided. Constrained
comboboxes use the [useComboboxInputConstrained](#usecomboboxinputconstrained) hook and often have
two separate `input` elements.

- user input - This is the visible input and it should contain user-friendly values. Calls to `.focus()`
  or `.blur()` are redirected to this input. Any prop passed to the `*.Input` component that is not
  directly related to forms will be passed here (i.e. `data-testid`, `aria-*`, etc).
- form input - This input is only visible to forms and the `value` will usually be server IDs. If the
  combobox options don't have an ID and only use user values, the value of this input will be the
  same as the user input. Any prop related to the function of forms will be passed here. For
  example, the `name` attribute will be passed here. The `ref` will be pointed to this element.

### Arbitrary

An arbitrary combobox allows the user to enter any value. The list of options are presented as
suggestions and selecting an option will prefill the combobox with the value of the option. The user
is still allowed to modify the combobox even after an option is entered. With arbitrary comboboxes,
there is only one `input` element. Arbitrary combobox inputs should use the
[useComboboxInputArbitrary](#usecomboboxinputarbirary) hook.

```

File: modules/react/combobox/lib/hooks/useComboboxKeyboardTypeAhead.ts (hunk 1 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/hooks/useComboboxKeyboardTypeAhead.ts:25]: SOFAR:

> _noun_
> 
> a system in which the sound waves from an underwater explosion are detected and located by three or more listening stations, useful in determining the position at sea of [survivors](https://www.google.com/search?sca_esv=e8b89075e9150def&sxsrf=ADLYWILxuesJAN14vw7rMUFV6xeoRVa9wg:1728871443452&q=survivors&si=ACC90nxMSPeZfdJJjQgDsdZJuFuJHqGJ1Cp6iSAb1ykqwEzDHU8Qu46wubqw7zqubJWF3Ccudp4QaDN3txWbuAboM4cn_KSKdY421fnb6YfouMdx_DyRplk%3D&expnd=1&sa=X&ved=2ahUKEwjr6f6B5IyJAxU3IDQIHUXMBqsQyecJegQIQhAN) of a disaster.

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx (hunk 3 above)
Comment (NicholasBoll) [reply modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:283]: Yes. That comment prompted a rewrite of the `Autocomplete` example.

**Pull request merged**